karoo_gp/base_class.py

```diff
@@ -172,63 +173,68 @@ def __init__(self, kernel='m', tree_type='r', tree_depth_base=3,
+        self.population = Population.generate(log=self.log,
+                                              pause=self.pause,
+                                              error=self.error,
+                                              tree_type=self.tree_type,
+                                              tree_depth_base=self.tree_depth_base,
+                                              tree_depth_max=tree_depth_max,
+                                              tree_pop_max=tree_pop_max,
+                                              functions=self.functions,
+                                              terminals=self.terminals,
+                                              rng=self.rng,
+                                              fitness_type=self.fitness_type)
+
+        # MAY REDESIGN - moved from init
+        self.log(f'\n\t\033[32m Press \033[36m\033[1m?\033[0;0m\033[32m at any '
+                 f'\033[36m\033[1m(pause)\033[0;0m\033[32m, or '
+                 f'\033[36m\033[1mENTER\033[0;0m \033[32mto continue the run\033[0;0m',
+             display=['i'])
+        self.pause(display=['i'])
+        self.population.evaluate(log=self.log,
+                                 pause=self.pause,
+                                 error=self.error,
+                                 data_train=self.data_train,
+                                 kernel=self.kernel,
+                                 data_train_rows=self.data_train_rows,
+                                 tf_device_log=self.tf_device_log,
+                                 class_labels=self.class_labels,
+                                 tf_device=self.tf_device,
+                                 terminals=self.terminals,
+                                 precision=self.precision,
+                                 savefile=self.savefile,
+                                 fx_data_tree_write=self.fx_data_tree_write)
         return
 
+
     def log(self, msg, display={'i', 'g', 'm', 'db'}):
-        show = self.display in display or display == 'all'
-        if show:
+        if self.display in display or display == 'all':
             print(msg)
 
     def pause(self, display={'i', 'g', 'm', 'db'}):
-        if (self.pause_callback and
-            (self.display in display or display == None)):
-            self.pause_callback()
+        if not self.pause_callback or not callable(self.pause_callback):
+            self.log('No pause callback function provided')
+            return
+        if self.display in display or display == None:
+            self.pause_callback(self)
 
-    def pause_callback(self):
-        self.fx_karoo_pause_refer()
+    def error(self, msg, display={'i', 'g', 'm', 'db'}):
+        self.log(msg, display)
+        self.pause(display)
 
 
     #+++++++++++++++++++++++++++++++++++++++++++++
     #   Methods to Run Karoo GP                  |
     #+++++++++++++++++++++++++++++++++++++++++++++
 
-    def fit(self):
+    def fit(self, X=None, y=None):
```

Comment:
Actually `X, y` is convention in sklearn. It got into this commit by accident (it was from some earlier experimentation), but it'll be used soon.

I think `X` is big because it's 2+ dimensional (multiple values for each data sample), and `y` is small because it's one-dimensional (a single target value for each sample).